`or { return }` C error for map when value has function type

### Describe the bug

When using a map with value type `fn` and trying to securely get a value from this map by key with `or { return }` construction, a C error happens.

### Expected Behavior

No error, correct value is returned for existed key and `return` is called for unexisted key.

### Current Behavior

```
/tmp/v_60000/code.404484527250648116.tmp.c:1260: error: 'main__handle_mouse_down_signal' undeclared
builder error: 
==================
C error. This should never happen.
```

### Reproduction Steps

```v
const mouse_action_to_function_map = {
	1: handle_mouse_down_signal
}

fn handle_mouse_down_signal() {
	
}

test := mouse_action_to_function_map[1] or { return }
```

### Possible Solution

_No response_

### Additional Information/Context

Without `or { return }` code works

### V version

V 0.4.0 7f8749a

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.0 7f8749a
OS: linux, "Garuda Linux"
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 3800X 8-Core Processor

getwd: /home/yuart
vexe: /home/yuart/Projects/v/v
vexe mtime: 2023-07-05 15:00:34

vroot: OK, value: /home/yuart/Projects/v
VMODULES: OK, value: /home/yuart/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.41.0
Git vroot status: weekly.2023.27-18-g7f8749af
.git/config present: true

CC version: cc (GCC) 13.1.1 20230429
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1065,9 +1065,12 @@ fn (mut g Gen) expr_string_surround(prepend string, expr ast.Expr, append string
 // all unified in one place so that it doesnt break
 // if one location changes
 fn (mut g Gen) option_type_name(t ast.Type) (string, string) {
-	base := g.base_type(t)
+	mut base := g.base_type(t)
 	mut styp := ''
 	sym := g.table.sym(t)
+	if sym.info is ast.FnType {
+		base = 'anon_fn_${g.table.fn_type_signature(sym.info.func)}'
+	}
 	if sym.language == .c && sym.kind == .struct_ {
 		styp = '${c.option_name}_${base.replace(' ', '_')}'
 	} else {
@@ -1087,6 +1090,9 @@ fn (mut g Gen) result_type_name(t ast.Type) (string, string) {
 	}
 	mut styp := ''
 	sym := g.table.sym(t)
+	if sym.info is ast.FnType {
+		base = 'anon_fn_${g.table.fn_type_signature(sym.info.func)}'
+	}
 	if sym.language == .c && sym.kind == .struct_ {
 		styp = '${c.result_name}_${base.replace(' ', '_')}'
 	} else {

--- a/vlib/v/tests/option_map_fn_type_value_test.v
+++ b/vlib/v/tests/option_map_fn_type_value_test.v
@@ -1,0 +1,13 @@
+const mouse_action_to_function_map = {
+	1: handle_mouse_down_signal
+}
+
+fn handle_mouse_down_signal() string {
+	return 'hello'
+}
+
+fn test_option_map_fn_type_value() {
+	t := mouse_action_to_function_map[1] or { return }
+	println(t())
+	assert t() == 'hello'
+}

